dev/archery/archery/docker.py

```diff
@@ -0,0 +1,152 @@
+            self.config = yaml.load(fp)
+
+        self.nodes = dict(flatten(self.config['x-hierarchy']))
+        self.dotenv = dotenv_values(self.dotenv_path)
+        self.bin = default_bin(compose_bin, "docker-compose")
+
+    def validate(self):
+        services = self.config['services'].keys()
+        nodes = self.nodes.keys()
+        errors = []
+
+        for name in nodes - services:
+            errors.append(
+                'Service `{}` is defined in `x-hierarchy` bot not in '
+                '`services`'.format(name)
+            )
+        for name in services - nodes:
+            errors.append(
+                'Service `{}` is defined in `services` but not in '
+                '`x-hierarchy`'.format(name)
+            )
+
+        # trigger docker-compose's own validation
+        result = self._execute('ps', check=False, stderr=subprocess.PIPE,
+                               stdout=subprocess.PIPE)
+
+        if result.returncode != 0:
+            # strip the intro line of docker-compose errors
+            errors += result.stderr.decode().splitlines()[1:]
+
+        if errors:
+            msg = '\n'.join([' - {}'.format(msg) for msg in errors])
+            raise ValueError(
+                'Found errors with docker-compose:\n{}'.format(msg)
+            )
+
+    def _compose_env(self):
+        return dict(os.environ, **self.dotenv)
+
+    def _validate_image(self, name):
+        if name not in self.nodes:
+            raise UndefinedImage(name)
+
+    def _execute(self, *args, **kwargs):
+        # set default arguments for docker-compose
+        return super().run('--file', str(self.config_path), *args, **kwargs)
+
+    def build(self, image, cache=True, cache_leaf=True):
+        self._validate_image(image)
+        env = self._compose_env()
+
+        # build all parents
+        for parent in self.nodes[image]:
+            if cache:
+                self._execute('pull', '--ignore-pull-failures', parent,
+                              env=env)
+                self._execute('build', parent, env=env)
+            else:
+                self._execute('build', '--no-cache', parent, env=env)
+
+        # build the image at last
+        if cache and cache_leaf:
+            self._execute('pull', '--ignore-pull-failures', image, env=env)
+            self._execute('build', image, env=env)
+        else:
+            self._execute('build', '--no-cache', image, env=env)
+
+    def run(self, image, command=None, env=None):
+        self._validate_image(image)
+        args = []
+        if env is not None:
+            for k, v in env.items():
+                args.extend(['-e', '{}={}'.format(k, v)])
+        args.append(image)
+        if command is not None:
+            args.append(command)
+        self._execute('run', '--rm', *args, env=self._compose_env())
+
+    def push(self, image, user, password):
+        try:
+            Docker().run('login', '-u', user, '-p', password)
```

Comment:
Use `--password-stdin`, otherwise they'll leak in the process argv via ps/top/etc and you won't need the next exception re-raising.

Reply:
Created follow-up https://issues.apache.org/jira/browse/ARROW-8698